File: client/src/services/validate.js
import { apiUrl } from './api';

/**
 * Send image to backend for validation.
 * @param {File} file
 * @returns {Promise<{valid, format, width, height, size_bytes, message}>}
 */
export async function validateImage(file) {
  const fd = new FormData();
  fd.append('file', file, file.name);

  const res = await fetch(apiUrl('/api/validate'), {
    method: 'POST',
    body: fd,
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: 'Validation failed' }));
    throw new Error(err.detail || 'Validation failed');
  }

  return res.json();
}

File: src-tauri/src/commands/validate.rs
/// POST /api/validate
/// Validates an uploaded image and returns format, dimensions, size.

use axum::{
    extract::Multipart,
    http::StatusCode,
    response::Json,
};
use serde_json::{json, Value};

use crate::services::validator::validate_image;

pub async fn handle_validate(mut multipart: Multipart) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let (data, filename) = extract_file(&mut multipart).await?;

    let result = validate_image(&data, &filename);

    Ok(Json(json!({
        "valid":      result.valid,
        "format":     result.format,
        "width":      result.width,
        "height":     result.height,
        "size_bytes": result.size_bytes,
        "message":    result.message,
    })))
}

/// Extract the `file` field from multipart, returning (bytes, filename).
pub async fn extract_file(
    multipart: &mut Multipart,
) -> Result<(Vec<u8>, String), (StatusCode, Json<Value>)> {
    while let Ok(Some(field)) = multipart.next_field().await {
        let name = field.name().unwrap_or("").to_string();
        if name == "file" {
            let filename = field
                .file_name()
                .unwrap_or("upload")
                .to_string();
            let data = field.bytes().await.map_err(|e| {
                (
                    StatusCode::BAD_REQUEST,
                    Json(json!({ "detail": format!("Failed to read file: {}", e) })),
                )
            })?;
            return Ok((data.to_vec(), filename));
        }
    }

    Err((
        StatusCode::BAD_REQUEST,
        Json(json!({ "detail": "No file field found in request" })),
    ))
}